src/Workspaces/Core/Portable/Workspace/Solution/Checksum_Factory.cs

```diff
@@ -23,47 +22,100 @@ internal partial class Checksum
         // https://github.com/dotnet/runtime/blob/f2db6d6093c54e5eeb9db2d8dcbe15b2db92ad8c/src/libraries/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/SHA256.cs#L18-L19
         private const int SHA256HashSizeBytes = 256 / 8;
 
+#if NET5_0_OR_GREATER
         private static readonly ObjectPool<IncrementalHash> s_incrementalHashPool =
             new(() => IncrementalHash.CreateHash(HashAlgorithmName.SHA256), size: 20);
+#else
+        private static readonly ObjectPool<SHA256> s_incrementalHashPool =
+            new(SHA256.Create, size: 20);
+#endif
 
+#if !NET5_0_OR_GREATER
         // Dedicated pools for the byte[]s we use to create checksums from two or three existing checksums. Sized to
         // exactly the space needed to splat the existing checksum data into the array and then hash it.
 
         private static readonly ObjectPool<byte[]> s_twoChecksumByteArrayPool = new(() => new byte[HashSize * 2]);
         private static readonly ObjectPool<byte[]> s_threeChecksumByteArrayPool = new(() => new byte[HashSize * 3]);
+#endif
 
         public static Checksum Create(IEnumerable<string> values)
         {
+#if NET5_0_OR_GREATER
+            using var pooledHash = s_incrementalHashPool.GetPooledObject();
+
+            foreach (var value in values)
+            {
+                pooledHash.Object.AppendData(MemoryMarshal.AsBytes(value.AsSpan()));
+                pooledHash.Object.AppendData(MemoryMarshal.AsBytes("\0".AsSpan()));
```

Comment:
unrelated to your change, but it's def weird we append these null bytes.

Reply:
➡️ Ensures the hash for `[ "Text1", "" ]` is different from the hash of `[ "Text", "1" ]`.

Reply:
Fair. I originally felt that that wasn't necessary, but then remembered these checksums very much *are* expected to be unique keys to data, so this is actually very important.  Thanks!